package egovframework.hyb.add.cmr.service.impl;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.sql.Timestamp;
import java.text.SimpleDateFormat;
import java.util.Locale;

import javax.annotation.Resource;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import egovframework.com.cmm.security.DeviceAPIFileUploadValidator;
import egovframework.hyb.add.cmr.service.CameraAndroidAPIFileVO;
import egovframework.hyb.add.cmr.service.EgovCameraAndroidAPIService;
import egovframework.rte.fdl.cmmn.EgovAbstractServiceImpl;
import egovframework.rte.fdl.cmmn.exception.BaseRuntimeException;
import egovframework.rte.fdl.cmmn.exception.FdlException;
import egovframework.rte.fdl.idgnr.EgovIdGnrService;
import egovframework.rte.fdl.property.EgovPropertyService;

/**  
 * @Class Name : EgovFileTransfer.java
 * @Description : EgovFileTransfer
 * @
 * @  수정일                 수정자                 수정내용
 * @ ---------   ---------   -------------------------------
 * @ 2012. 7. 10.  서준식                  최초생성
 * @ 2012. 7. 23.  이율경                  커스터마이징
 * @ 2017.02.27    최두영        시큐어코딩(ES)-36. 부적절한 예외 처리[CWE253, CWE-440, CWE-754]
 *  
 * @author 디바이스 API 실행환경 개발팀
 * @since 2012. 7. 10.
 * @version 1.0
 * @see
 * 
 *  Copyright (C) by MOPAS All right reserved.
 */
@Service("EgovCameraAndroidFileMngUtil")
public class EgovCameraAndroidFileMngUtil extends EgovAbstractServiceImpl {
    
    /** EgovCameraAndroidAPIService */
    @Resource(name = "EgovCameraAndroidAPIService")
    private EgovCameraAndroidAPIService egovCameraAndroidAPIService;

    private static final Logger LOGGER = LoggerFactory.getLogger(EgovCameraAndroidFileMngUtil.class);
    
    /** propertiesService */
    @Resource(name = "propertiesService")
    protected EgovPropertyService propertiesService;
    
    @Resource(name="egovFileIdGnrService")
    private EgovIdGnrService egovFileIdGnrService;
    
    
    public CameraAndroidAPIFileVO writeUploadedFile(MultipartFile file) {
        
        DeviceAPIFileUploadValidator.validateCmrUpload(file);

        String originFileName = file.getOriginalFilename();
        LOGGER.debug("originFileName={}", originFileName);
        int index = originFileName.lastIndexOf(".");
        String fileExt = originFileName.substring(index + 1);
        String newName = "IMAGE_" + getTimeStamp() + ".jpg";
        
        String filePath = propertiesService.getString("fileStorePath");
        
        CameraAndroidAPIFileVO fileVO = new CameraAndroidAPIFileVO();
        
        try {
			fileVO.setFileSn(egovFileIdGnrService.getNextIntegerId());
		} catch (FdlException e) {
			throw new BaseRuntimeException(e);
		}
        fileVO.setFileStreCours(filePath);
        fileVO.setStreFileNm(newName);
        fileVO.setOrignlFileNm(originFileName);
        fileVO.setFileExtsn(fileExt);
        fileVO.setFileSize(Long.toString(file.getSize()));
        
        IOException excep = null;
        
        if(!file.isEmpty()){
            InputStream input = null;
            FileOutputStream out = null;
            try {
                byte[] bytes = file.getBytes();
                input = new ByteArrayInputStream(bytes);
                
                File videoFile = new File(filePath + newName);
                out = new FileOutputStream(videoFile);
                int nextChar;
                while((nextChar = input.read()) != -1){
                    out.write(nextChar);
                    out.flush();                    
                }
            //2017-02-27 최두영 시큐어코딩(ES)-36. 부적절한 예외 처리[CWE253, CWE-440, CWE-754] 95-95
            }catch(IOException e){
            	throw new BaseRuntimeException("Fail to open FileOutPutStream : ", e);
            }finally{
                try {
                    if(out != null){
                        out.close();
                    }
                } catch(IOException e) {
                	throw new BaseRuntimeException(e);
                }
            }
        }

        egovCameraAndroidAPIService.insertCameraPhotoAlbumFile(fileVO);
        
        return fileVO;
    }
    
    private static String getTimeStamp(){
        String rtnStr = null;

        // 문자열로 변환하기 위한 패턴 설정(년도-월-일 시:분:초:초(자정이후 초))
        String pattern = "yyyyMMddhhmmssSSS";

        SimpleDateFormat sdfCurrent = new SimpleDateFormat(pattern, Locale.KOREA);
        Timestamp ts = new Timestamp(System.currentTimeMillis());
        rtnStr = sdfCurrent.format(ts.getTime());

        return rtnStr;
    }
}
